pkg/importer/imageio-datasource.go

```diff
@@ -196,26 +219,70 @@ func (is *ImageioDataSource) pollProgress(reader *util.CountingReader, idleTime,
 	}
 }
 
-func createImageioReader(ctx context.Context, ep string, accessKey string, secKey string, certDir string, diskID string) (io.ReadCloser, uint64, *ovirtsdk4.ImageTransfer, ConnectionInterface, error) {
+// IsDeltaCopy is called to determine if this is a full copy or one delta copy stage
+// in a multi-stage migration.
+func (is *ImageioDataSource) IsDeltaCopy() bool {
+	result := is.previousSnapshot != "" && is.currentSnapshot != ""
+	return result
+}
+
+func createImageioReader(ctx context.Context, ep string, accessKey string, secKey string, certDir string, diskID string, currentCheckpoint string, previousCheckpoint string) (io.ReadCloser, uint64, *ovirtsdk4.ImageTransfer, ConnectionInterface, error) {
 	conn, err := newOvirtClientFunc(ep, accessKey, secKey)
 	if err != nil {
 		return nil, uint64(0), nil, conn, errors.Wrap(err, "Error creating connection")
 	}
 
-	it, total, err := getTransfer(conn, diskID)
+	// Get disk
+	disk, err := getDisk(conn, diskID)
+	if err != nil {
+		return nil, uint64(0), nil, conn, err
+	}
+
+	// Get snapshot if a checkpoint was specified
+	var snapshot *ovirtsdk4.DiskSnapshot
+	if currentCheckpoint != "" {
+		var snapshotErr error
+		snapshot, snapshotErr = getSnapshot(conn, disk, currentCheckpoint)
+		if snapshot == nil { // Snapshot not found, check for a disk with a matching image ID
+			imageID, available := disk.ImageId()
+			if !available {
+				return nil, uint64(0), nil, conn, errors.Wrap(snapshotErr, "Could not get disk's image ID!")
+			}
+			if imageID != currentCheckpoint {
+				return nil, uint64(0), nil, conn, errors.Wrapf(snapshotErr, "Snapshot %s not found!", currentCheckpoint)
+			}
+			// Matching ID: use disk as checkpoint
+			klog.Infof("Snapshot ID %s found on disk %s, transferring active disk as checkpoint.", currentCheckpoint, diskID)
+		}
+	}
+
+	// For regular imports and the first stage of a multi-stage import, download as raw.
+	// For actual snapshots and active disks whose image ID has been specified as the
+	// snapshot to import, download as QCOW.
+	formatType := ovirtsdk4.DISKFORMAT_RAW
+	if currentCheckpoint != "" && previousCheckpoint != "" {
+		klog.Info("Downloading snapshot as qcow")
+		formatType = ovirtsdk4.DISKFORMAT_COW
+	}
+
+	// Get transfer ticket for disk or snapshot
+	it, total, err := getTransfer(conn, disk, snapshot, formatType)
 	if err != nil {
 		return nil, uint64(0), it, conn, err
 	}
+	if it == nil {
+		return nil, uint64(0), nil, conn, errors.New("returned ImageTransfer was nil")
+	}
 
 	// Use the create client from http source.
 	client, err := createHTTPClient(certDir)
 	if err != nil {
-		cancelTransfer(conn, it)
+		cleanupTransfer(conn, it)
```

Comment:
Would it make sense to defer the cleanup at the top of this function so we ensure it always get called regardless of errors. Also do we need to clean up the transfer if we are successful? If so, I would also in that defer at least check the err returned during cleanupTransfer, and at a minimum log it

Reply:
No, at this point the transfer ticket is opened, but no data has been copied. It only calls cleanupTransfer on the error exits to try to make sure the disk does not stay locked after failing. This definitely should not clean up the ticket on success, so the defer would have to have some way of knowing about each failure anyway.

The cleanupTransfer function itself already does a good amount of logging, and I wouldn't want an error from cleanupTransfer to obscure the original error that caused it to get called in the first place.